File: app/src/main/java/com/robifr/ledger/ui/create_product/CreateProductName.java
/**
 * Copyright (c) 2022-present Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.create_product;

import android.text.Editable;
import android.widget.EditText;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.robifr.ledger.ui.EditTextWatcher;
import java.util.Objects;

public class CreateProductName {
  @NonNull private final CreateProductFragment _fragment;
  @NonNull private final NameTextWatcher _nameTextWatcher;

  public CreateProductName(@NonNull CreateProductFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
    this._nameTextWatcher = new NameTextWatcher(this._fragment.fragmentBinding().name);

    this._fragment.fragmentBinding().name.addTextChangedListener(this._nameTextWatcher);
  }

  public void setInputtedNameText(@Nullable String name) {
    final String currentText = this._fragment.fragmentBinding().name.getText().toString();
    if (currentText.equals(name)) return;

    final int cursorPosition = name != null ? name.length() : 0;

    // Remove listener to prevent any sort of formatting although there isn't.
    this._fragment.fragmentBinding().name.removeTextChangedListener(this._nameTextWatcher);
    this._fragment.fragmentBinding().name.setText(name);
    this._fragment.fragmentBinding().name.setSelection(cursorPosition);
    this._fragment.fragmentBinding().name.addTextChangedListener(this._nameTextWatcher);
  }

  public void setError(@NonNull String message, boolean isEnabled) {
    Objects.requireNonNull(message);

    this._fragment.fragmentBinding().nameLayout.setError(message);
    this._fragment.fragmentBinding().nameLayout.setErrorEnabled(isEnabled);
  }

  private class NameTextWatcher extends EditTextWatcher {
    public NameTextWatcher(@NonNull EditText view) {
      super(view);
    }

    @Override
    public void afterTextChanged(@NonNull Editable editable) {
      super.afterTextChanged(editable);
      CreateProductName.this._fragment.createProductViewModel().onNameTextChanged(this.newText());
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/createproduct/CreateProductName.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createproduct

import android.text.Editable
import com.google.android.material.textfield.TextInputEditText
import com.robifr.ledger.ui.EditTextWatcher
import com.robifr.ledger.ui.createproduct.viewmodel.CreateProductViewModel

class CreateProductName(private val _fragment: CreateProductFragment) {
  private val _nameTextWatcher: NameTextWatcher =
      NameTextWatcher(_fragment.createProductViewModel, _fragment.fragmentBinding.name)

  init {
    _fragment.fragmentBinding.name.addTextChangedListener(_nameTextWatcher)
  }

  fun setInputtedNameText(name: String, errorMessage: String?) {
    if (_fragment.fragmentBinding.name.getText().toString() != name) {
      // Remove listener to prevent any sort of formatting although there isn't.
      _fragment.fragmentBinding.name.removeTextChangedListener(_nameTextWatcher)
      _fragment.fragmentBinding.name.setText(name)
      _fragment.fragmentBinding.name.setSelection(name.length)
      _fragment.fragmentBinding.name.addTextChangedListener(_nameTextWatcher)
    }
    _fragment.fragmentBinding.nameLayout.setError(errorMessage)
  }
}

private class NameTextWatcher(
    private val _createProductViewModel: CreateProductViewModel,
    editText: TextInputEditText
) : EditTextWatcher(editText) {
  override fun afterTextChanged(editable: Editable) {
    super.afterTextChanged(editable)
    _createProductViewModel.onNameTextChanged(newText())
  }
}
